roachtest: django failed

roachtest.django [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10689032?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10689032?buildTab=artifacts#/django) on master @ [0b0a212ecbe514b55301f4a963edc92b16de9bab](https://github.com/cockroachdb/cockroach/commits/0b0a212ecbe514b55301f4a963edc92b16de9bab):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.2.0-alpha.00000000-dev-0b0a212ecbe514b55301f4a963edc92b16de9bab
Tests run against django cockroach-4.1.x
9951 Total Tests Run
9949 tests passed
2 tests failed
639 tests skipped
2 tests ignored
0 tests passed unexpectedly
2 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: queries.tests.Queries1Tests.test_ticket2306 (unexpected)
--- FAIL: queries.tests.Queries1Tests.test_ticket4464 (unexpected)
--- SKIP: backends.mysql.test_creation.DatabaseCreationTests.test_create_test_db_unexpected_error due to MySQL tests (expected)
For a full summary look at the django artifacts 
An updated blocklist (djangoBlocklist) is available in the artifacts' django log
test artifacts and logs in: /artifacts/django/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*django.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29116

```
======================================================================
FAIL: test_ticket2306 (queries.tests.Queries1Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/data1/django/tests/queries/tests.py", line 228, in test_ticket2306
    self.assertSequenceEqual(
AssertionError: Sequences differ: <QuerySet []> != [<Item: one>]

Second sequence contains 1 additional elements.
First extra element 0:
<Item: one>

- <QuerySet []>
+ [<Item: one>]

======================================================================
FAIL: test_ticket4464 (queries.tests.Queries1Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/data1/django/tests/queries/tests.py", line 259, in test_ticket4464
    self.assertSequenceEqual(
AssertionError: Sequences differ: <QuerySet []> != [<Item: one>]

Second sequence contains 1 additional elements.
First extra element 0:
<Item: one>

- <QuerySet []>
+ [<Item: one>]

----------------------------------------------------------------------
```

This might be a regression. @andyyang890 can you see if this reproduces? If so, we should find the query it's running, and potentially hand this off to another team.

roachtest.django [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10722908?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10722908?buildTab=artifacts#/django) on master @ [4a614f89cea81bf94674d6072c3bbf30502244d4](https://github.com/cockroachdb/cockroach/commits/4a614f89cea81bf94674d6072c3bbf30502244d4):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.2.0-alpha.00000000-dev-4a614f89cea81bf94674d6072c3bbf30502244d4
Tests run against django cockroach-4.1.x
9951 Total Tests Run
9949 tests passed
2 tests failed
639 tests skipped
2 tests ignored
0 tests passed unexpectedly
2 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: queries.tests.Queries1Tests.test_ticket2306 (unexpected)
--- FAIL: queries.tests.Queries1Tests.test_ticket4464 (unexpected)
--- SKIP: backends.mysql.test_creation.DatabaseCreationTests.test_create_test_db_unexpected_error due to MySQL tests (expected)
For a full summary look at the django artifacts 
An updated blocklist (djangoBlocklist) is available in the artifacts' django log
test artifacts and logs in: /artifacts/django/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_fs=ext4</code>
, <code>ROACHTEST_localSSD=true</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*django.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.django [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10735768?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10735768?buildTab=artifacts#/django) on master @ [84a5961ebde5898d3c97aa97da59719c12a5ed16](https://github.com/cockroachdb/cockroach/commits/84a5961ebde5898d3c97aa97da59719c12a5ed16):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.2.0-alpha.00000000-dev-84a5961ebde5898d3c97aa97da59719c12a5ed16
Tests run against django cockroach-4.1.x
9951 Total Tests Run
9949 tests passed
2 tests failed
639 tests skipped
2 tests ignored
0 tests passed unexpectedly
2 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: queries.tests.Queries1Tests.test_ticket2306 (unexpected)
--- SKIP: backends.mysql.test_creation.DatabaseCreationTests.test_create_test_db_unexpected_error due to MySQL tests (expected)
--- FAIL: queries.tests.Queries1Tests.test_ticket4464 (unexpected)
For a full summary look at the django artifacts 
An updated blocklist (djangoBlocklist) is available in the artifacts' django log
test artifacts and logs in: /artifacts/django/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_fs=ext4</code>
, <code>ROACHTEST_localSSD=true</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*django.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.django [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10750288?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10750288?buildTab=artifacts#/django) on master @ [df82b5306e945d396269f28874d98fa567f4ac32](https://github.com/cockroachdb/cockroach/commits/df82b5306e945d396269f28874d98fa567f4ac32):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.2.0-alpha.00000000-dev-df82b5306e945d396269f28874d98fa567f4ac32
Tests run against django cockroach-4.1.x
9951 Total Tests Run
9949 tests passed
2 tests failed
639 tests skipped
2 tests ignored
0 tests passed unexpectedly
2 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: queries.tests.Queries1Tests.test_ticket4464 (unexpected)
--- SKIP: backends.mysql.test_creation.DatabaseCreationTests.test_create_test_db_unexpected_error due to MySQL tests (expected)
--- FAIL: queries.tests.Queries1Tests.test_ticket2306 (unexpected)
For a full summary look at the django artifacts 
An updated blocklist (djangoBlocklist) is available in the artifacts' django log
test artifacts and logs in: /artifacts/django/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*django.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.django [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10756585?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10756585?buildTab=artifacts#/django) on master @ [aacba20d325e5702836e9a76be646b5f1bd922af](https://github.com/cockroachdb/cockroach/commits/aacba20d325e5702836e9a76be646b5f1bd922af):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.2.0-alpha.00000000-dev-aacba20d325e5702836e9a76be646b5f1bd922af
Tests run against django cockroach-4.1.x
9951 Total Tests Run
9949 tests passed
2 tests failed
639 tests skipped
2 tests ignored
0 tests passed unexpectedly
2 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: queries.tests.Queries1Tests.test_ticket2306 (unexpected)
--- SKIP: backends.mysql.test_creation.DatabaseCreationTests.test_create_test_db_unexpected_error due to MySQL tests (expected)
--- FAIL: queries.tests.Queries1Tests.test_ticket4464 (unexpected)
For a full summary look at the django artifacts 
An updated blocklist (djangoBlocklist) is available in the artifacts' django log
test artifacts and logs in: /artifacts/django/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*django.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.django [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10763018?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10763018?buildTab=artifacts#/django) on master @ [aacba20d325e5702836e9a76be646b5f1bd922af](https://github.com/cockroachdb/cockroach/commits/aacba20d325e5702836e9a76be646b5f1bd922af):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.2.0-alpha.00000000-dev-aacba20d325e5702836e9a76be646b5f1bd922af
Tests run against django cockroach-4.1.x
9951 Total Tests Run
9949 tests passed
2 tests failed
639 tests skipped
2 tests ignored
0 tests passed unexpectedly
2 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: queries.tests.Queries1Tests.test_ticket2306 (unexpected)
--- SKIP: backends.mysql.test_creation.DatabaseCreationTests.test_create_test_db_unexpected_error due to MySQL tests (expected)
--- FAIL: queries.tests.Queries1Tests.test_ticket4464 (unexpected)
For a full summary look at the django artifacts 
An updated blocklist (djangoBlocklist) is available in the artifacts' django log
test artifacts and logs in: /artifacts/django/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*django.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.django [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10771334?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10771334?buildTab=artifacts#/django) on master @ [428dc9da6a320de218460de6c6c8807caa4ded98](https://github.com/cockroachdb/cockroach/commits/428dc9da6a320de218460de6c6c8807caa4ded98):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.2.0-alpha.00000000-dev-428dc9da6a320de218460de6c6c8807caa4ded98
Tests run against django cockroach-4.1.x
9951 Total Tests Run
9949 tests passed
2 tests failed
639 tests skipped
2 tests ignored
0 tests passed unexpectedly
2 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- SKIP: backends.mysql.test_creation.DatabaseCreationTests.test_create_test_db_unexpected_error due to MySQL tests (expected)
--- FAIL: queries.tests.Queries1Tests.test_ticket2306 (unexpected)
--- FAIL: queries.tests.Queries1Tests.test_ticket4464 (unexpected)
For a full summary look at the django artifacts 
An updated blocklist (djangoBlocklist) is available in the artifacts' django log
test artifacts and logs in: /artifacts/django/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*django.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.django [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10780210?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10780210?buildTab=artifacts#/django) on master @ [34699bb9c1557fce449e08a68cd259efec94926f](https://github.com/cockroachdb/cockroach/commits/34699bb9c1557fce449e08a68cd259efec94926f):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.2.0-alpha.00000000-dev-34699bb9c1557fce449e08a68cd259efec94926f
Tests run against django cockroach-4.1.x
9951 Total Tests Run
9949 tests passed
2 tests failed
639 tests skipped
2 tests ignored
0 tests passed unexpectedly
2 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: queries.tests.Queries1Tests.test_ticket2306 (unexpected)
--- FAIL: queries.tests.Queries1Tests.test_ticket4464 (unexpected)
--- SKIP: backends.mysql.test_creation.DatabaseCreationTests.test_create_test_db_unexpected_error due to MySQL tests (expected)
For a full summary look at the django artifacts 
An updated blocklist (djangoBlocklist) is available in the artifacts' django log
test artifacts and logs in: /artifacts/django/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*django.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/join
+++ b/pkg/sql/logictest/testdata/logic_test/join
@@ -1152,3 +1152,56 @@ SELECT abcd.*, dxby.* FROM abcd INNER JOIN dxby USING (d, b)
 ----
 a  b  c  d  d  x  b  y
 2  2  2  2  2  2  2  2
+
+# Regression test for #105608 - don't filter out NULL values from the
+# task-patient LEFT join.
+statement ok
+CREATE TABLE patient (id INT8 NOT NULL, site_id INT8, PRIMARY KEY (id));
+
+statement ok
+CREATE TABLE site (id INT8 NOT NULL, PRIMARY KEY (id));
+
+statement ok
+CREATE TABLE task (id INT8 NOT NULL, description VARCHAR(255), patient_id INT8, PRIMARY KEY (id));
+
+statement ok
+ALTER TABLE IF EXISTS patient ADD CONSTRAINT fkhty4ykfvf29xscswwd63mgdoc FOREIGN KEY (site_id) REFERENCES site;
+
+statement ok
+ALTER TABLE IF EXISTS task ADD CONSTRAINT fkkk2ow88d08vqxqyvvvmys1j2m FOREIGN KEY (patient_id) REFERENCES patient;
+
+statement ok
+INSERT INTO site(id) VALUES (1);
+INSERT INTO site(id) VALUES (2);
+INSERT INTO site(id) VALUES (3);
+
+statement ok
+INSERT INTO patient(site_id, id) VALUES (NULL, 4);
+INSERT INTO patient(site_id, id) VALUES (1, 5);
+INSERT INTO patient(site_id, id) VALUES (1, 6);
+INSERT INTO patient(site_id, id) VALUES (2, 7);
+INSERT INTO patient(site_id, id) VALUES (3, 8);
+
+statement ok
+INSERT INTO task(description, patient_id, id) VALUES ('taskWithoutPatient', NULL, 9);
+INSERT INTO task(description, patient_id, id) VALUES ('taskWithPatientWithoutSite', 4, 10);
+INSERT INTO task(description, patient_id, id) VALUES ('taskWithPatient1WithValidSite1', 5, 11);
+INSERT INTO task(description, patient_id, id) VALUES ('taskWithPatient2WithValidSite1', 6, 12);
+INSERT INTO task(description, patient_id, id) VALUES ('taskWithPatient3WithValidSite2', 7, 13);
+INSERT INTO task(description, patient_id, id) VALUES ('taskWithPatientWithInvalidSite', 8, 14);
+
+query ITI rowsort
+SELECT task0_.id AS id1_2_, task0_.description AS descript2_2_, task0_.patient_id AS patient_3_2_
+FROM task AS task0_ WHERE EXISTS (
+	SELECT task1_.id FROM task AS task1_
+	LEFT JOIN patient AS patient2_
+	ON task1_.patient_id = patient2_.id
+	LEFT JOIN site AS site3_
+	ON patient2_.site_id = site3_.id
+	WHERE (task1_.id = task0_.id) AND ((patient2_.id IS NULL) OR (site3_.id IN (2, 1)))
+);
+----
+9   taskWithoutPatient              NULL
+11  taskWithPatient1WithValidSite1  5
+12  taskWithPatient2WithValidSite1  6
+13  taskWithPatient3WithValidSite2  7

--- a/pkg/sql/opt/xform/join_order_builder.go
+++ b/pkg/sql/opt/xform/join_order_builder.go
@@ -472,6 +472,11 @@ func (jb *JoinOrderBuilder) ensureClosure(join memo.RelExpr) {
 	for col, ok := reps.Next(0); ok; col, ok = reps.Next(col + 1) {
 		// Get all columns which are known to be equal to this column.
 		equivGroup := equivFDs.ComputeEquivGroup(col)
+		if !equivGroup.Intersects(join.Relational().NotNullCols) {
+			// If the equivalent columns are nullable, it would be invalid to add an
+			// equality edge because SQL equality rejects NULL values.
+			continue
+		}
 
 		// Ensure that there exists an edge for each distinct pair of equivalent
 		// columns.

--- a/pkg/sql/opt/xform/testdata/rules/join_order
+++ b/pkg/sql/opt/xform/testdata/rules/join_order
@@ -3156,3 +3156,216 @@ project
  │         └── t1.c = t2.c
  └── projections
       └── 1
+
+# Regression test for #105608 - don't infer an equality on nullable columns from
+# the top-level join's functional dependencies.
+exec-ddl
+CREATE TABLE patient (id INT8 NOT NULL, site_id INT8, PRIMARY KEY (id));
+----
+
+exec-ddl
+CREATE TABLE site (id INT8 NOT NULL, PRIMARY KEY (id));
+----
+
+exec-ddl
+CREATE TABLE task (id INT8 NOT NULL, description VARCHAR(255), patient_id INT8, PRIMARY KEY (id));
+----
+
+exec-ddl
+ALTER TABLE IF EXISTS patient ADD CONSTRAINT fkhty4ykfvf29xscswwd63mgdoc FOREIGN KEY (site_id) REFERENCES site;
+----
+
+exec-ddl
+ALTER TABLE IF EXISTS task ADD CONSTRAINT fkkk2ow88d08vqxqyvvvmys1j2m FOREIGN KEY (patient_id) REFERENCES patient;
+----
+
+# The edge "task0_.patient_id = task1_.patient_id" should not be created.
+reorderjoins
+SELECT task0_.id AS id1_2_, task0_.description AS descript2_2_, task0_.patient_id AS patient_3_2_
+FROM task AS task0_ WHERE EXISTS (
+	SELECT task1_.id FROM task AS task1_
+	LEFT JOIN patient AS patient2_
+	ON task1_.patient_id = patient2_.id
+	LEFT JOIN site AS site3_
+	ON patient2_.site_id = site3_.id
+	WHERE (task1_.id = task0_.id) AND ((patient2_.id IS NULL) OR (site3_.id IN (2, 1)))
+);
+----
+--------------------------------------------------------------------------------
+Join Tree #1
+--------------------------------------------------------------------------------
+  left-join (hash)
+   ├── scan task [as=task1_]
+   ├── scan patient [as=patient2_]
+   └── filters
+        └── task1_.patient_id = patient2_.id
+Vertexes
+  A:
+    scan task [as=task1_]
+  B:
+    scan patient [as=patient2_]
+Edges
+  task1_.patient_id = patient2_.id [left, ses=AB, tes=AB, rules=()]
+Joining AB
+  A B [left, refs=AB]
+Joins Considered: 1
+--------------------------------------------------------------------------------
+Join Tree #2
+--------------------------------------------------------------------------------
+  left-join (hash)
+   ├── left-join (hash)
+   │    ├── scan task [as=task1_]
+   │    ├── scan patient [as=patient2_]
+   │    └── filters
+   │         └── task1_.patient_id = patient2_.id
+   ├── scan site [as=site3_]
+   └── filters
+        └── site_id = site3_.id
+Vertexes
+  A:
+    scan task [as=task1_]
+  B:
+    scan patient [as=patient2_]
+  C:
+    scan site [as=site3_]
+Edges
+  task1_.patient_id = patient2_.id [left, ses=AB, tes=AB, rules=()]
+  site_id = site3_.id [left, ses=BC, tes=BC, rules=()]
+Joining AB
+  A B [left, refs=AB]
+Joining BC
+  B C [left, refs=BC]
+Joining ABC
+  A BC [left, refs=AB]
+  AB C [left, refs=BC]
+Joins Considered: 4
+--------------------------------------------------------------------------------
+Join Tree #3
+--------------------------------------------------------------------------------
+  semi-join (hash)
+   ├── scan task [as=task0_]
+   ├── select
+   │    ├── left-join (hash)
+   │    │    ├── left-join (hash)
+   │    │    │    ├── scan task [as=task1_]
+   │    │    │    ├── scan patient [as=patient2_]
+   │    │    │    └── filters
+   │    │    │         └── task1_.patient_id = patient2_.id
+   │    │    ├── scan site [as=site3_]
+   │    │    └── filters
+   │    │         └── site_id = site3_.id
+   │    └── filters
+   │         └── (patient2_.id IS NULL) OR (site3_.id IN (1, 2))
+   └── filters
+        └── task1_.id = task0_.id
+Vertexes
+  D:
+    scan task [as=task0_]
+  E:
+    select
+     ├── left-join (hash)
+     │    ├── left-join (hash)
+     │    │    ├── scan task [as=task1_]
+     │    │    ├── scan patient [as=patient2_]
+     │    │    └── filters
+     │    │         └── task1_.patient_id = patient2_.id
+     │    ├── scan site [as=site3_]
+     │    └── filters
+     │         └── site_id = site3_.id
+     └── filters
+          └── (patient2_.id IS NULL) OR (site3_.id IN (1, 2))
+Edges
+  task1_.id = task0_.id [semi, ses=DE, tes=DE, rules=()]
+Joining DE
+  D E [semi, refs=DE]
+Joins Considered: 1
+--------------------------------------------------------------------------------
+Join Tree #4
+--------------------------------------------------------------------------------
+  inner-join (hash)
+   ├── scan task [as=task0_]
+   ├── select
+   │    ├── left-join (hash)
+   │    │    ├── left-join (hash)
+   │    │    │    ├── scan task [as=task1_]
+   │    │    │    ├── scan patient [as=patient2_]
+   │    │    │    └── filters
+   │    │    │         └── task1_.patient_id = patient2_.id
+   │    │    ├── scan site [as=site3_]
+   │    │    └── filters
+   │    │         └── site_id = site3_.id
+   │    └── filters
+   │         └── (patient2_.id IS NULL) OR (site3_.id IN (1, 2))
+   └── filters
+        └── task1_.id = task0_.id
+Vertexes
+  D:
+    scan task [as=task0_]
+  E:
+    select
+     ├── left-join (hash)
+     │    ├── left-join (hash)
+     │    │    ├── scan task [as=task1_]
+     │    │    ├── scan patient [as=patient2_]
+     │    │    └── filters
+     │    │         └── task1_.patient_id = patient2_.id
+     │    ├── scan site [as=site3_]
+     │    └── filters
+     │         └── site_id = site3_.id
+     └── filters
+          └── (patient2_.id IS NULL) OR (site3_.id IN (1, 2))
+Edges
+  task1_.id = task0_.id [inner, ses=DE, tes=DE, rules=()]
+Joining DE
+  D E [inner, refs=DE]
+  E D [inner, refs=DE]
+Joins Considered: 2
+================================================================================
+Final Plan
+================================================================================
+project
+ ├── columns: id1_2_:1!null descript2_2_:2 patient_3_2_:3
+ ├── key: (1)
+ ├── fd: (1)-->(2,3)
+ └── project
+      ├── columns: task0_.id:1!null task0_.description:2 task0_.patient_id:3 task1_.id:6!null
+      ├── key: (6)
+      ├── fd: (1)-->(2,3), (6)-->(3), (1)==(6), (6)==(1)
+      └── inner-join (lookup task [as=task0_])
+           ├── columns: task0_.id:1!null task0_.description:2 task0_.patient_id:3 task1_.id:6!null task1_.patient_id:8 patient2_.id:11 site_id:12 site3_.id:15
+           ├── key columns: [6] = [1]
+           ├── lookup columns are key
+           ├── key: (6)
+           ├── fd: (1)-->(2,3), (6)-->(8,11,12,15), (11)-->(12), (1)==(6), (6)==(1), (3)==(8), (8)==(3)
+           ├── select
+           │    ├── columns: task1_.id:6!null task1_.patient_id:8 patient2_.id:11 site_id:12 site3_.id:15
+           │    ├── key: (6)
+           │    ├── fd: (6)-->(8,11,12,15), (11)-->(12)
+           │    ├── left-join (hash)
+           │    │    ├── columns: task1_.id:6!null task1_.patient_id:8 patient2_.id:11 site_id:12 site3_.id:15
+           │    │    ├── multiplicity: left-rows(exactly-one), right-rows(zero-or-more)
+           │    │    ├── key: (6)
+           │    │    ├── fd: (6)-->(8,11,12,15), (11)-->(12)
+           │    │    ├── left-join (hash)
+           │    │    │    ├── columns: task1_.id:6!null task1_.patient_id:8 patient2_.id:11 site_id:12
+           │    │    │    ├── multiplicity: left-rows(exactly-one), right-rows(zero-or-more)
+           │    │    │    ├── key: (6)
+           │    │    │    ├── fd: (6)-->(8,11,12), (11)-->(12)
+           │    │    │    ├── scan task [as=task1_]
+           │    │    │    │    ├── columns: task1_.id:6!null task1_.patient_id:8
+           │    │    │    │    ├── key: (6)
+           │    │    │    │    └── fd: (6)-->(8)
+           │    │    │    ├── scan patient [as=patient2_]
+           │    │    │    │    ├── columns: patient2_.id:11!null site_id:12
+           │    │    │    │    ├── key: (11)
+           │    │    │    │    └── fd: (11)-->(12)
+           │    │    │    └── filters
+           │    │    │         └── task1_.patient_id:8 = patient2_.id:11 [outer=(8,11), constraints=(/8: (/NULL - ]; /11: (/NULL - ]), fd=(8)==(11), (11)==(8)]
+           │    │    ├── scan site [as=site3_]
+           │    │    │    ├── columns: site3_.id:15!null
+           │    │    │    └── key: (15)
+           │    │    └── filters
+           │    │         └── site_id:12 = site3_.id:15 [outer=(12,15), constraints=(/12: (/NULL - ]; /15: (/NULL - ]), fd=(12)==(15), (15)==(12)]
+           │    └── filters
+           │         └── (patient2_.id:11 IS NULL) OR (site3_.id:15 IN (1, 2)) [outer=(11,15)]
+           └── filters (true)

